### PR TITLE
check expected grpc error

### LINKED ISSUE #400: Seeing warnings in ztunnel logs that could cause unnecessary concerns
Happening every 30 mins:

2023-02-15T20:54:40.204428Z  WARN xds{id=1}: ztunnel::xds::client: XDS client error: gRPC error (Unknown error): error reading a body from connection: broken pipe, retrying

2023-02-15T21:24:40.813192Z  WARN xds{id=2}: ztunnel::xds::client: XDS client error: gRPC error (Unknown error): error reading a body from connection: broken pipe, retrying

@howardjohn mentioned this is ztunnel disconnects with istiod every 30 mins so istiod can rebalance and we hide this with Envoy.

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fix #400

### CODE REVIEW

**Comment by bleggett** [comment]
This looks like it fixes https://github.com/istio/ztunnel/issues/400?

**Review by bleggett — COMMENTED** [review]
Admittedly it's a small change so if you don't want to that's fine - but since this is sort of a behavioral quirk, describing/fencing the quirk with a unit test would be good.

**Review by bleggett — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
>@hzxuzhonghu This looks like it fixes https://github.com/istio/ztunnel/issues/400?

Yes

**Review by ymesika — APPROVED** [review]
LGTM

File: src/xds/client.rs [hunk 1]
```diff
@@ -266,6 +265,26 @@ impl AdsClient {
                 tokio::time::sleep(backoff).await;
                 backoff
             }
+            Err(e @ Error::GrpcStatus(_)) => {
+                let err_detail = e.to_string();
+                let status: tonic::Status = e.try_into().unwrap();
```
Comment (howardjohn) [inline src/xds/client.rs:270]: unwrap is not needed here, we can do:
```
            Err(ref e @ Error::GrpcStatus(ref status)) => {
                let err_detail = e.to_string();
```

and remove the TryInto impl above

File: src/xds/client.rs [hunk 2]
```diff
@@ -266,6 +265,26 @@ impl AdsClient {
...
+                if status.code() == tonic::Code::Unknown
```
Comment (howardjohn) [inline src/xds/client.rs:271]: why Unknown? we don't have that in the Go version and it seems important

File: src/xds/client.rs [hunk 3]
```diff
@@ -266,6 +265,26 @@ impl AdsClient {
...
+                        && status.message().contains("transport is closing"))
+                {
+                    debug!("XDS client terminated: {}, retrying", err_detail);
+                    self.metrics
+                        .increment(&ConnectionTerminationReason::Reconnect);
+                } else {
+                    warn!("XDS client error: {}, retrying", err_detail);
+                    self.metrics.increment(&ConnectionTerminationReason::Error);
+                }
+                // For gRPC errors, we connect immediately
```
Comment (howardjohn) [inline src/xds/client.rs:284]: Its not clear to me why we don't want backoff here but we do for Connection. Does envoy do the same?

Comment (hzxuzhonghu) [reply src/xds/client.rs:284]: This is just same as previous, i do not change it.  

@ramaraochavali Do you know envoy's behavior

Comment (ramaraochavali) [reply src/xds/client.rs:284]: Envoy there is a back off https://github.com/envoyproxy/envoy/blob/a4053922d7311e11c18e764ff5e5bfa8f71cd334/source/common/config/grpc_stream.h#L21 and there are efforts to make it configurable https://github.com/envoyproxy/envoy/pull/24701

File: src/xds/client.rs (hunk 2 above)
Comment (hzxuzhonghu) [reply src/xds/client.rs:271]: https://github.com/istio/ztunnel/issues/400  unknown error here, not sure the error in  rust is same as go 

```
2023-02-15T20:54:40.204428Z WARN xds{id=1}: ztunnel::xds::client: XDS client error: gRPC error (Unknown error): error reading a body from connection: broken pipe, retrying```

File: src/xds/client.rs (hunk 3 above)
Comment (hzxuzhonghu) [reply src/xds/client.rs:283]: Thanks， suggest open a new issue to tack that

Comment (hzxuzhonghu) [reply src/xds/client.rs:283]: https://github.com/istio/ztunnel/issues/413

File: src/xds/client.rs (hunk 1 above)
Comment (hzxuzhonghu) [reply src/xds/client.rs]: good , i didn't know ref usage

**Review by howardjohn — APPROVED** [review]

**Pull request merged**